### PR TITLE
Metro mode selector trip form

### PR DESCRIPTION
**Pull request by daniel-heppner-ibigroup**
This PR includes the actual trip-form component for the new mode selector.

### CODE REVIEW

File: packages/trip-form/src/MetroModeSelector/index.tsx [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+import styled, { css } from "styled-components";
+
+import SubSettingsPane, { defaultMessages } from "./SubSettingsPane";
+
+const invisibleCss = css`
+  clip: rect(0, 0, 0, 0);
+  height: 0;
+  overflow: hidden;
+  width: 0;
+  position: absolute;
```
Comment (binh-dam-ibigroup) [inline packages/trip-form/src/MetroModeSelector/index.tsx:28]: Sort CSS.

File: packages/trip-form/src/MetroModeSelector/index.tsx [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+  padding: 0;
+
+  /* <legend> is not shown visually. */
+  & > legend {
+    ${invisibleCss}
+  }
+`;
+
+const accentColor = "#084c8d";
+const activeHoverColor = "0e5faa";
```
Comment (binh-dam-ibigroup) [inline packages/trip-form/src/MetroModeSelector/index.tsx:50]: Did I forget a `#` sign here?

File: packages/trip-form/src/MetroModeSelector/index.tsx [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+      </button>
+      {renderDropdown && (
+        <FloatingFocusManager context={context}>
+          <HoverPanel
+            // This library relies on prop spreading
+            // eslint-disable-next-line react/jsx-props-no-spreading
+            {...getFloatingProps()}
+            // Matches ID on Header element in SubSettingsPane
+            aria-labelledby={`metro-mode-selector-${modeButton.key}-button-label`}
+            ref={floating}
```
Comment (binh-dam-ibigroup) [inline packages/trip-form/src/MetroModeSelector/index.tsx:280]: This one should be fixed: When mouse-hovering out of a label that has the popup open, the focus is set on that label, even if keyboard focus was on a different element before.

File: packages/trip-form/src/MetroModeSelector/MetroModeSelector.story.tsx [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+      // eslint-disable-next-line react/jsx-props-no-spreading
+      {...args}
+    />
+  </QueryParamProvider>
+);
+
+export default {
+  component: MetroModeSelectorComponent,
+  title: "Trip Form Components/Metro Mode Selector",
+  argTypes: {
```
Comment (binh-dam-ibigroup) [inline packages/trip-form/src/MetroModeSelector/MetroModeSelector.story.tsx:113]: Sort attributes

File: packages/trip-form/src/MetroModeSelector/utils.ts [hunk 5]
```diff
@@ -0,0 +1,308 @@
+import { useState } from "react";
+import {
+  useQueryParam,
+  DelimitedArrayParam,
+  ObjectParam
+} from "use-query-params";
+import { QueryParamConfig, decodeQueryParams } from "serialize-query-params";
```
Comment (binh-dam-ibigroup) [inline packages/trip-form/src/MetroModeSelector/utils.ts:7]: Sort imports by package name

File: packages/trip-form/src/MetroModeSelector/utils.ts [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+import {
+  ModeButtonDefinition,
+  ModeSetting,
+  ModeSettingValues,
+  TransportMode
+} from "@opentripplanner/types";
+
+import { TRANSIT_SUBMODES_AND_TRANSIT } from "@opentripplanner/core-utils/lib/query-gen";
```
Comment (binh-dam-ibigroup) [inline packages/trip-form/src/MetroModeSelector/utils.ts:15]: Was the default export from `query-gen` exported in `index` in core-utils?

File: packages/trip-form/src/MetroModeSelector/MetroModeSelector.story.tsx [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+    onToggleModeButton(key);
+  };
+
+  const setModeSettingValueAction = (event: QueryParamChangeEvent) => {
+    setModeSettingValue(event);
+    onSetModeSettingValue(event);
+  };
+
+  return (
+    <div style={{ position: "relative" }}>
```
Comment (binh-dam-ibigroup) [inline packages/trip-form/src/MetroModeSelector/MetroModeSelector.story.tsx:86]: This surrounding `<div>` can be removed.

File: packages/trip-form/src/MetroModeSelector/MetroModeSelector.story.tsx [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+    modes: [{ mode: "CAR" }]
+  }
+];
+
+const modeSettingDefinitionsWithDropdown = [
+  ...modeSettingDefinitions,
+  {
+    applicableMode: "TRANSIT",
+    type: "DROPDOWN",
+    key: "busColor",
```
Comment (binh-dam-ibigroup) [inline packages/trip-form/src/MetroModeSelector/MetroModeSelector.story.tsx:43]: Sort attributes

File: packages/trip-form/src/__tests__/mode-selector-utils.js [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+      {
+        mode: "RAIL"
+      }
+    ],
+    enabled: true,
+    modeSettings: [
+      {
+        applicableMode: "TRANSIT",
+        type: "CHECKBOX",
+        key: "wheelchair",
```
Comment (binh-dam-ibigroup) [inline packages/trip-form/src/__tests__/mode-selector-utils.js:28]: Try to sort attributes.

File: packages/trip-form/src/styled.ts [hunk 10]
```diff
@@ -13,10 +13,10 @@ export const SettingsHeader = styled.div``;
 export const SettingsSection = styled.div``;
 
 export const SettingLabel = styled.label`
+  font-weight: 400;
   padding-left: 6px;
-
   svg {
-    width: 22pt;
+    width: 22pt;
```
Comment (binh-dam-ibigroup) [inline packages/trip-form/src/styled.ts:19]: Revert this change.

File: packages/trip-form/package.json [hunk 11]
```diff
@@ -10,19 +10,22 @@
   "types": "lib/index.d.ts",
   "private": false,
   "dependencies": {
-    "@opentripplanner/core-utils": "^8.1.1",
+    "@floating-ui/react": "^0.19.2",
+    "@opentripplanner/core-utils": "^9.0.0-alpha.7",
```
Comment (binh-dam-ibigroup) [inline packages/trip-form/package.json:14]: Use the non-alpha version number.

**Review by binh-dam-ibigroup — CHANGES_REQUESTED** [review]
Main thing is the keyboard focus being stolen. See also some cleanup items.

File: packages/trip-form/src/MetroModeSelector/index.tsx (hunk 3 above)
Comment (binh-dam-ibigroup) [reply packages/trip-form/src/MetroModeSelector/index.tsx:280]: Could you revert https://github.com/opentripplanner/otp-ui/pull/555/commits/ab66a4db693cf65c0ab3dfd96e8946475852204f? I think that change has more side effects than the initial behavior.

Comment (binh-dam-ibigroup) [reply packages/trip-form/src/MetroModeSelector/index.tsx:280]: I'm leaning towards disabling hovering for the duration that a keyboard-triggered popup is open. I think we could do that in a separate PR but let me run that with @demory.

Comment (binh-dam-ibigroup) [reply packages/trip-form/src/MetroModeSelector/index.tsx:280]: @daniel-heppner-ibigroup See https://github.com/ibi-group/otp-ui/pull/16 that addresses the issues above.

**Comment by daniel-heppner-ibigroup** [comment]
@binh-dam-ibigroup We re-added the `fillModeIcon` option since SoundTransit wants to keep their circle icons and we cannot fill them. It seems like the `fill: currentcolor` CSS property is needed for Houston Metro's icons to look correct.

**Comment by binh-dam-ibigroup** [comment]
> @binh-dam-ibigroup We re-added the `fillModeIcon` option since SoundTransit wants to keep their circle icons and we cannot fill them. It seems like the `fill: currentcolor` CSS property is needed for Houston Metro's icons to look correct.

@daniel-heppner-ibigroup If so, please add a story to illustrate that prop.

**Comment by binh-dam-ibigroup** [comment]
Gonna unassign myself, it seems there are more changes coming.

**Review by binh-dam-ibigroup — APPROVED** [review]
I recommend moving the "Bus color" setting out of the language files that are being shipped (I left those settings untranslated). Otherwise, with the clean up comments, it looks solid!

File: packages/trip-form/i18n/en-US.yml [hunk 12]
```diff
@@ -6,6 +6,30 @@ otpUi:
     depart: Depart at
     now: Leave now
     time: Time
+  ModeSelector:
+    labels:
+      BICYCLE: Bike
+      CAR: Drive
+      RENT: Rent
+      TRANSIT: Transit
```
Comment (miles-grant-ibigroup) [inline packages/trip-form/i18n/en-US.yml:14]: It looks like these might need to be lowercase? 
Screenshot 2023-03-30 at 12 12 34 PM

File: packages/trip-form/src/MetroModeSelector/index.tsx [hunk 13]
```diff
@@ -0,0 +1,407 @@
...
+  margin: 0 4px 0 0;
+  padding: 0;
+
+  /* <legend> is not shown visually. */
+  & > legend {
+    ${invisibleCss}
+  }
+`;
+
+const accentColor = "#084c8d";
```
Comment (miles-grant-ibigroup) [inline packages/trip-form/src/MetroModeSelector/index.tsx:49]: Can we make this configurable/importable from props?

File: packages/trip-form/src/MetroModeSelector/index.tsx [hunk 14]
```diff
@@ -0,0 +1,407 @@
...
+`;
+
+const HoverPanel = styled.div`
+  min-width: 300px;
+  padding: 0 10px;
+  width: 75%;
+  z-index: 10;
+`;
+
+const HoverInnerContainer = styled.div`
```
Comment (miles-grant-ibigroup) [inline packages/trip-form/src/MetroModeSelector/index.tsx:154]: I'd like to make some changes here -- the submode icons should ideally be separated a bit, and possibly get icons? 

I'd also like to see a bit more padding on the container

File: packages/trip-form/src/MetroModeSelector/index.tsx [hunk 15]
```diff
@@ -0,0 +1,407 @@
...
+  background: #fff;
+  border-radius: 4px;
+  box-shadow: 0 5px 15px rgba(0, 0, 0, 0.05), 0 4px 10px rgba(0, 123, 255, 0.25);
```
Comment (miles-grant-ibigroup) [inline packages/trip-form/src/MetroModeSelector/index.tsx:157]: This is a bit intense we should probably increase the spread and lower the opacity -- also can we have the color match the button color?

File: packages/trip-form/src/MetroModeSelector/index.tsx [hunk 16]
```diff
@@ -0,0 +1,407 @@
...
+  font-size: 90%;
+  font-weight: bold;
+  padding: 5px;
+  pointer-events: none;
+`;
+
+const Arrow = styled.div`
+  background: #fff;
+  box-shadow: 3px -2px 3px rgba(0, 0, 0, 0.06),
+    2px -2px 5px -3px rgba(0, 123, 255, 0.25);
```
Comment (miles-grant-ibigroup) [inline packages/trip-form/src/MetroModeSelector/index.tsx:168]: Again can we have this match the accent color? Also should this box shadow be extracted

File: packages/trip-form/src/MetroModeSelector/utils.ts [hunk 17]
```diff
@@ -0,0 +1,311 @@
...
+): ModeSetting[] {
+  return modeSettings.map(setting => {
+    const value = values[setting.key];
+    let convertedVal;
+    if (setting.type === "CHECKBOX") {
+      // If the parameter is anything besides the string "true", it will be false
+      convertedVal = value === "true" || value === true;
+    } else if (setting.type === "SLIDER") {
+      convertedVal = Number(value);
+    }
```
Comment (miles-grant-ibigroup) [inline packages/trip-form/src/MetroModeSelector/utils.ts:78]: This might be cleaner as a method that converts `setting.type` into a converted value, then that can bet set to a const

File: packages/trip-form/src/index.ts [hunk 18]
```diff
@@ -8,19 +8,29 @@ import SettingsSelectorPanel from "./SettingsSelectorPanel";
...
   CheckboxSelector,
   DateTimeSelector,
+  defaultModeSettings,
   DropdownSelector,
   GeneralSettingsPanel,
+  getActivatedModesFromQueryParams,
+  MetroModeSelector,
   ModeButton,
   ModeSelector,
   SettingsSelectorPanel,
```
Comment (miles-grant-ibigroup) [inline packages/trip-form/src/index.ts:29]: Is this actually backwards compatible with the `SettingSelectorPanel`? If not we can probably remove it

**Review by miles-grant-ibigroup — CHANGES_REQUESTED** [review]
Looking good! Mostly small changes, mostly around styling

File: packages/trip-form/src/MetroModeSelector/index.tsx (hunk 13 above)
Comment (binh-dam-ibigroup) [reply packages/trip-form/src/MetroModeSelector/index.tsx:49]: See https://github.com/ibi-group/configurations/pull/379/files#r1153514224 for an example on how to do this.

File: packages/trip-form/i18n/en-US.yml (hunk 12 above)
Comment (daniel-heppner-ibigroup) [reply packages/trip-form/i18n/en-US.yml:14]: Yes, you're right! Good catch. I made them lowercase and added a .toLowerCase() in the switch statement.

File: packages/trip-form/src/MetroModeSelector/utils.ts [hunk 19]
```diff
@@ -0,0 +1,311 @@
...
+  const [reactState, setReactState] = useState<Type>(defaultState);
+
+  if (!storeInQueryParam) {
+    return [reactState, setReactState];
+  }
+
+  if (qpState === undefined) {
+    return [defaultState, setQpState];
+  }
+  return [qpState, setQpState];
```
Comment (binh-dam-ibigroup) [inline packages/trip-form/src/MetroModeSelector/utils.ts:174]: Consider this for a fix PR: As written, if a top-level mode is not activated, the query parameters for that mode still appear in the URL bar. Consider removing query parameters from the URL for unactivated modes.

File: packages/trip-form/src/MetroModeSelector/index.tsx (hunk 13 above)
Comment (daniel-heppner-ibigroup) [reply packages/trip-form/src/MetroModeSelector/index.tsx:49]: Added some props for override

File: packages/trip-form/src/MetroModeSelector/index.tsx (hunk 14 above)
Comment (daniel-heppner-ibigroup) [reply packages/trip-form/src/MetroModeSelector/index.tsx:154]: I switched the fields to flexbox and added a gap. Is that what you wanted? Icon is a good idea, I'll add support for that.

File: packages/trip-form/src/MetroModeSelector/index.tsx (hunk 15 above)
Comment (daniel-heppner-ibigroup) [reply packages/trip-form/src/MetroModeSelector/index.tsx:157]: Tuned it down and made it match!

File: packages/trip-form/src/MetroModeSelector/index.tsx (hunk 16 above)
Comment (daniel-heppner-ibigroup) [reply packages/trip-form/src/MetroModeSelector/index.tsx:168]: Oh good idea

File: packages/trip-form/src/index.ts (hunk 18 above)
Comment (daniel-heppner-ibigroup) [reply packages/trip-form/src/index.ts:29]: Well, the OTP-UI PR is fully backwards compatible, it's OTP-RR side that breaks stuff. I think we can remove this in a future breaking change.

File: packages/trip-form/src/MetroModeSelector/index.tsx (hunk 14 above)
Comment (daniel-heppner-ibigroup) [reply packages/trip-form/src/MetroModeSelector/index.tsx:161]: image
now with cute little icons

**Review by miles-grant-ibigroup — CHANGES_REQUESTED** [review]
Another cosmetic pass, functionality wise this works great!

File: packages/trip-form/i18n/en-US.yml [hunk 20]
```diff
@@ -6,6 +6,30 @@ otpUi:
     depart: Depart at
     now: Leave now
     time: Time
+  ModeSelector:
+    labels:
+      bicycle: Bike
```
[the remark this answers is not in the document]
Comment (daniel-heppner-ibigroup) [reply packages/trip-form/i18n/en-US.yml:11]: Should be bicycle. Will make sure it's consistent

File: packages/trip-form/src/styled.ts [hunk 21]
```diff
@@ -13,11 +13,8 @@ export const SettingsHeader = styled.div``;
 export const SettingsSection = styled.div``;
 
 export const SettingLabel = styled.label`
+  font-weight: 400;
```
Comment (binh-dam-ibigroup) [inline packages/trip-form/src/styled.ts:16]: I'd write `font-weight: normal`.

**Review by binh-dam-ibigroup — APPROVED** [review]
Don't see anything that's broken, and it seems to be playing nice with OTP-RR 🤞.

**Review by miles-grant-ibigroup — APPROVED** [review]
Some small notes but for the most part I think this is ready for launch! We can always do more cleanup/removal of dead code later on

**Review by philip-cline — APPROVED** [review]
This looks good to me after some testing. Love that we can now deselect all modes, can't wait to use that!
